flocker/node/agents/ebs.py

```diff
@@ -243,59 +335,88 @@ def _blockdevicevolume_from_ebs_volume(ebs_volume):
     )
 
 
-def _wait_for_volume(volume,
-                     start_status,
-                     transient_status,
-                     end_status):
+def _get_ebs_volume_state(volume):
+    """
+    Fetch input EBS volume's latest state from backend.
+
+    :param boto.ec2.volume volume: Volume that needs state update.
+
+    :returns: EBS volume with latest state known to backend.
+    :rtype: boto.ec2.volume
+
+    """
+    return volume.update()
+
+
+def _wait_for_volume_state_change(operation,
+                                  volume,
+                                  update=_get_ebs_volume_state,
+                                  timeout=VOLUME_STATE_CHANGE_TIMEOUT):
```

Comment:
This makes some progress towards what I was thinking of, but not the whole way.

Here's the logic I was going for. Instead of:

```
def wait(x, y, get_update):
    start = time.time()
    while time.time() - start < TIMEOUT:
        if x and y and get_update().foo:
            return
     raise TimeoutError()
```

We split this into two:
```
def should_finish(x, y, foo, start_time, current_time):
     if current_time - start_time < T:
         raise TimeoutError()
     return x and y and foo

def wait(x, y, get_update):
    start = time.time()
    while True:
        if should_finish(x, y, get_update().foo, start, time.time()):
            return
```

Notice that all the decision logic branches are now in `should_finish`. At the same time `should_finish` is completely side-effect free ("purely functional"): no calls to the outside universe, so you can test it all in isolation.